src/coord/src/coord.rs

```diff
@@ -4096,6 +4111,121 @@ impl Coordinator {
                 .expect("sending to internal_cmd_tx cannot fail");
         });
     }
+
+    /// Validate that sources which only support a single materialization are only materialized once
+    fn validate_single_materializations(
+        &self,
+        index_to_build: Option<GlobalId>,
+        index_view: View,
+        expr: OptimizedMirRelationExpr,
+    ) -> Result<(), CoordError> {
+        let id = if let Some(id) = index_to_build {
+            id
+        } else {
+            return Ok(());
+        };
+        let imports = self.catalog.get_imports::<usize>(&id, &expr, None);
+        let mut require_single = vec![];
+        self.deps_require_single(imports.iter().map(|idx| idx.dep_id), &mut require_single);
+        if !require_single.is_empty() {
+            for idx in self.catalog.entries().filter_map(|e| e.index()) {
+                if index_view.depends_on.contains(&idx.on) {
+                    continue;
+                }
+                let on = self.catalog.get_by_id(&idx.on);
+
+                self.reuses_single(on, &require_single)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    fn validate_idx_materializations(&self, depends_on: &[GlobalId]) -> Result<(), CoordError> {
+        let mut require_single = Vec::new();
+        self.deps_require_single(depends_on.iter().copied(), &mut require_single);
+
+        if !require_single.is_empty() {
+            for idx in self.catalog.entries().filter_map(|e| e.index()) {
+                let on = self.catalog.get_by_id(&idx.on);
+
+                self.reuses_single(on, &require_single)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Recurse through all dependencies finding sources which require single materialization
+    ///
+    /// `required` is populated with all the dependencies that require a single materialization
+    fn deps_require_single(
+        &self,
+        ids: impl Iterator<Item = GlobalId>,
```

Comment:
nit: argument position impl traits are a bit limited in that the type can't be specified at the call site if needed. I would use a named generic parameter